core/src/main/scala/kafka/admin/ZkSecurityMigrator.scala

```diff
@@ -244,4 +135,119 @@ object ZkSecurityMigrator extends Logging {
           e.printStackTrace()
     }
   }
-}
\ No newline at end of file
+}
+
+class ZkSecurityMigrator(zkUtils: ZkUtils) extends Logging {
+  private val workQueue = new LinkedBlockingQueue[Runnable]
+  private implicit val threadExecutionContext =
+    ExecutionContext.fromExecutor(new ThreadPoolExecutor(1,
+      Runtime.getRuntime().availableProcessors(),
+      5000,
+      TimeUnit.MILLISECONDS,
+      workQueue))
+
+  private val futures = new ArrayBuffer[Future[String]]
+
+  private def setAclsRecursively(path: String) = {
+    info("Setting ACL for path %s".format(path))
+    val setPromise = Promise[String]
+    val childrenPromise = Promise[String]
+    futures.synchronized {
+      futures += setPromise.future
+      futures += childrenPromise.future
+    }
+    zkUtils.zkConnection.getZookeeper.setACL(path, ZkUtils.DefaultAcls(zkUtils.isSecure), -1, SetACLCallback, setPromise)
+    zkUtils.zkConnection.getZookeeper.getChildren(path, false, GetChildrenCallback, childrenPromise)
+  }
+
+  private object GetChildrenCallback extends ChildrenCallback {
+    def processResult(rc: Int,
+                      path: String,
+                      ctx: Object,
+                      children: java.util.List[String]) {
+      //val list = Option(children).map(_.asScala)
+      val zkHandle = zkUtils.zkConnection.getZookeeper
+      val promise = ctx.asInstanceOf[Promise[String]]
+      Code.get(rc) match {
+        case Code.OK =>
+          // Set ACL for each child
+          Future {
+            for (child <- children.asScala)
+              setAclsRecursively(s"$path/$child")
```

Comment:
I think this does the same as the `StringBuilder` and it's much clearer. As far as I can see, this is not one of the cases where StringBuilder is actually a big performance win.